runtime/doc/ui.txt

```diff
@@ -0,0 +1,284 @@
+["popupmenu_show", items, selected, row, col]
+	`items` is an array of the items to show, the
+	items are themselves arrays of the form [word, kind, menu, info]
+	as defined at |complete-items|, except that `word` is replaced by
+	`abbr` if present.  `selected` is the initially selected item, either a
+	zero-based index into the array of items, or -1 if no item is
+	selected. `row` and `col` is the anchor position, where the first
+	character of the completed word will be.
+
+["popupmenu_select", selected]
+	An item in the currently displayed popupmenu is selected. `selected`
+	is either a zero-based index into the array of items from the last
+	`popupmenu_show` event, or -1 if no item is selected.
+
+["popupmenu_hide"]
+	The popupmenu is hidden.
+
+==============================================================================
+Tabline Events							   *ui-tabline*
+
+Only sent if `ext_tabline` option is set in |ui-options|
+
+["tabline_update", curtab, tabs]
+	Tabline was updated. UIs should present this data in a custom tabline
+	widget.
+	curtab:	  Current Tabpage
+	tabs:	  List of Dicts [{ "tab": Tabpage, "name": String }, ...]
+
+==============================================================================
+Cmdline Events							   *ui-cmdline*
+
+Only sent if `ext_cmdline` option is set in |ui-options|
+
+["cmdline_show", content, pos, firstc, prompt, indent, level]
+        content: List of [attrs, string]
+	         [[{}, "t"], [attrs, "est"], ...]
+	The `content` is the full content that should be displayed in the
+	cmdline, and the `pos` is the position of the cursor that in the
+	cmdline. This event will be triggered when you type in the cmdline.
+	The content is divided into chunks with different highlight attributes
+	represented as a dict, see |ui-event-highlight_set|.
+
+	`firstc` and `prompt` are text, that if non-empty should be
+	displayed in front of the command line. `firstc` always indicates
+	built in command lines such as `:` (ex command) and `/` `?` (search),
+	while `prompt` is an |input()| prompt. `indent` tells how many spaces
+	the content should be indented
+
+	The command line in nvim can be invoked recursively, for instance by
+	typing `<c-r>=` at the command line prompt. The `level` field is used
+	to distinguish different command lines active at the same time. The
+	first invoked command line will always have level 1, and the next
+	recursively invoked prompt will have level 2. Also, a command line
+	invoked while in command line window will have a higher level than
+	than the edited command line.
+
+["cmdline_pos", pos, level]
+	Change the cursor position in the cmdline.
+
+["cmdline_special_char", c, shift, level]
+	Display a special char in the cmdline at the cursor position. This is
+	typically used to indicate a pending state, like right after Ctrl-V.
+	if shift is true the text after the cursor should be shifted,
+	otherwise it should overshadow the char at the cursor.
+	Should be hidden at next cmdline_char
```

Comment:
@bfredl should cmdline_char be cmdline_pos?

(Just let me know; I'll fix it during merge)